src/modules/musig/keyagg_impl.h

```diff
@@ -0,0 +1,271 @@
+    ptr += 64;
+    secp256k1_fe_get_b32(ptr, &cache_i->second_pk_x);
+    ptr += 32;
+    memcpy(ptr, cache_i->pk_hash, 32);
+    ptr += 32;
+    *ptr = cache_i->internal_key_parity;
+    ptr += 1;
+    secp256k1_scalar_get_b32(ptr, &cache_i->tweak);
+}
+
+static int secp256k1_keyagg_cache_load(const secp256k1_context* ctx, secp256k1_keyagg_cache_internal *cache_i, const secp256k1_musig_keyagg_cache *cache) {
+    const unsigned char *ptr = cache->data;
+    ARG_CHECK(secp256k1_memcmp_var(ptr, secp256k1_musig_keyagg_cache_magic, 4) == 0);
+    ptr += 4;
+    secp256k1_point_load(&cache_i->pk, ptr);
+    ptr += 64;
+    secp256k1_fe_set_b32(&cache_i->second_pk_x, ptr);
+    ptr += 32;
+    memcpy(cache_i->pk_hash, ptr, 32);
+    ptr += 32;
+    cache_i->internal_key_parity = *ptr & 1;
+    ptr += 1;
+    secp256k1_scalar_set_b32(&cache_i->tweak, ptr, NULL);
+    return 1;
+}
+
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("KeyAgg list")||SHA256("KeyAgg list"). */
+static void secp256k1_musig_keyagglist_sha256(secp256k1_sha256 *sha) {
+    secp256k1_sha256_initialize(sha);
+
+    sha->s[0] = 0xb399d5e0ul;
+    sha->s[1] = 0xc8fff302ul;
+    sha->s[2] = 0x6badac71ul;
+    sha->s[3] = 0x07c5b7f1ul;
+    sha->s[4] = 0x9701e2eful;
+    sha->s[5] = 0x2a72ecf8ul;
+    sha->s[6] = 0x201a4c7bul;
+    sha->s[7] = 0xab148a38ul;
+    sha->bytes = 64;
+}
+
+/* Computes pk_hash = SHA256(pk[0], ..., pk[np-1]) */
+static int secp256k1_musig_compute_pk_hash(const secp256k1_context *ctx, unsigned char *pk_hash, const secp256k1_xonly_pubkey * const* pk, size_t np) {
+    secp256k1_sha256 sha;
+    size_t i;
+
+    secp256k1_musig_keyagglist_sha256(&sha);
+    for (i = 0; i < np; i++) {
+        unsigned char ser[32];
+        if (!secp256k1_xonly_pubkey_serialize(ctx, ser, pk[i])) {
+            return 0;
+        }
+        secp256k1_sha256_write(&sha, ser, 32);
+    }
+    secp256k1_sha256_finalize(&sha, pk_hash);
+    return 1;
+}
+
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("KeyAgg coefficient")||SHA256("KeyAgg coefficient"). */
+static void secp256k1_musig_keyaggcoef_sha256(secp256k1_sha256 *sha) {
+    secp256k1_sha256_initialize(sha);
+
+    sha->s[0] = 0x6ef02c5aul;
+    sha->s[1] = 0x06a480deul;
+    sha->s[2] = 0x1f298665ul;
+    sha->s[3] = 0x1d1134f2ul;
+    sha->s[4] = 0x56a0b063ul;
+    sha->s[5] = 0x52da4147ul;
+    sha->s[6] = 0xf280d9d4ul;
+    sha->s[7] = 0x4484be15ul;
+    sha->bytes = 64;
+}
+
+/* Compute KeyAgg coefficient which is constant 1 for the second pubkey and
+ * SHA256(pk_hash, x) where pk_hash is the hash of public keys otherwise. second_pk_x
+ * can be 0 in case there is no second_pk. Assumes both field elements x and
+ * second_pk_x are normalized. */
+static void secp256k1_musig_keyaggcoef_internal(secp256k1_scalar *r, const unsigned char *pk_hash, const secp256k1_fe *x, const secp256k1_fe *second_pk_x) {
```

Comment:
If we implemented this correctly then it shouldn't cause problems because that's what is proven secure ([Appendix B](https://eprint.iacr.org/2020/1261.pdf)